### ISSUE TITLE
Package Relay Project Tracking

### ISSUE THREAD

**Issue by glozow** [bitcoin/bitcoin#27463]
### Please describe the feature you'd like to see added.

**Mempool, Policy, Validation**

- [x] #22290
- [ ] Avoid the risk of below-minrelaytxfee transactions hanging around forever in the mempool (currently only possible through reorgs and replacements after package CPFP). There are 3 options:
   - Don't ever allow v1 and v2 txs below minrelaytxfee (#26933)
   - Mine everything that makes it into in the mempool (default blockmintxfee to 0). Trim descendant packages <=0 fees, even when the mempool is not full (#27018)
   - Other long-term improvements that address this problem
- [ ] When loading mempool, keep transactions that are below minimum feerate(s) but are CPFP'd. (#26711)
- [ ] Properly assess incentive compatibility, submit transactions with their in-package ancestor sets (#26711)
- [ ] Handle any arbitrary package, even if unsorted (#26711)
- [ ] Relax from child-with-parents to tx-with-ancestors (#26711)
- (safe to expose on P2P)
- [ ] Add v3 rules that make transactions/packages easier to work with (#25038)
- [ ] Introduce a Miner Incentive Compatibility Score rule to RBF (#25038)
- [ ] Enable Package RBF for v3 transactions (#25038)
- [ ] Add Ephemeral Anchors rules for v3 transactions (#26403)

**P2P, Net Processing**
(Rough outline. Will fill in when PR is open.)
- [ ] BIP 331 (bitcoin/bips/pull/1382)
- [ ] Distinguish between tx failures that {can, cannot} be different in a package. Add a separate rejections cache.
- [ ] Enable more reliable usage of orphanage for ancestor package relay.
- [ ] Negotiate package relay support with peers, gated by config flag (`sendpackages`)
- [ ] Upon receiving an orphan, request a list of its ancestors (`ancpkginfo`) instead of parents by txid. Use that to download missing ancestors.
- [ ] Using package info, download (`getpkgtxns`) and submit packages together.
- [ ] Avoid telling legacy peers about transactions that need package relay to be accepted, to save bandwidth.
- [ ] Default the config flag to true.

**Comment by instagibbs** [comment]
Very helpful. I think this helps define "sprints" that can be achieved to hit next milestones.

**Comment by ariard** [comment]
I think one more sub-category to track is the feedback collected on how package
relay/nversion=3 are fitting multi-party applications and contracting protocols
- Lightning: legacy channels / anchor output / taproot types
- Peerswap: claim with preimage + refund cooperatively + refund after csv passes
- Collaborative transaction: splicing + dual funding flows

The [0-conf flow](https://github.com/lightning/bolts/pull/910) of Lightning can be also considered. As deployed today by some LSPs where both incoming/outgoing HTLCs are allowed during the pending confirmation period, there is a risk in case of mempool congestions of the inbound funding transaction not confirming before the timelock of the offered HTLC output expires. The current nversion=3 size is 1000 vb which might be too low if LSPs would like to do batching of 0-confs. Note, I think all the pinning risks are not the same for 0-conf funding transactions than for time-sensitive confirmation of LN commitment+HTLCs transactions.

Mentioning 0-conf, and as raised on [#27643](https://github.com/bitcoin/bitcoin/pull/26525#issuecomment-1473067644) at some point in the development cycle of nversion=3 it might be good communication practice to publish than all 0confs users might have to upgrade their mempool monitoring stuff to reject nversion=3, or chain of transactions with them, as otherwise genuine upgrade to the version of Core with package relay might lower their security.

For Lightning, there is the question if the p2p package format is flexible enough for the Lightning channels to apply "new mempool rules" on "old in-flight channnel type". Like backward compatible effect of the mempools rules without off-chain channels having to cooperate on-resigning transactions. I believe that would require a Lightning node to sign the p2p package with the same pubkey than used with the channel (though I start to think we might have multiple types of backward/forward compatible application of mempools rules like the ones mentioned in [#25038](https://github.com/bitcoin/bitcoin/pull/25038#issuecomment-1405944407), like there is some matrix ?)

For the second-layers listed, I'm just considering the ones which have a) a mature specification and b) some substantial deployment though should we consider more second-layers or multi-party applications like vaults (non-deployed) ?

From a reviewer viewpoint, I think it could be very valuable if you can highlight the 2/3 PRs or mail posts you consider as top of the stack for reviews ? Thanks for a package-relay context-tracking meta-issue.

(Feel free to integrate what you find relevant of this comment to ease context tracking of package relay as a project)

**Comment by ajtowns** [comment]
I think at a high level, the "package relay" part looks like:

 - [x] make package mempool acceptance available via rpc in regtest
 - [ ] fix up package acceptance/retention/mining policies to be free from DoS vectors and to behave sensibly
 - [ ] behave sensibly with arbitrary txs in a package
 - [ ] implement bip331 to expose package relay over p2p not just rpc

And relatedly:

 - [ ] version 3 acceptance/rbf policies to limit pinnability
 - [ ] ephemeral anchors

Particularly if #26933 is applied, is there any reason not to make `submitpackage` available on all chains? It should behave no differently to submitting the transactions individually with a non-full mempool, no?

**Comment by ajtowns** [comment]
FWIW, I don't think "incentive compatibility" is a good description of our goal here -- I think there's three goals we want:

 - for most normal use cases (including newly invented ones) just submitting your txs over p2p should work fine
   - (otherwise people will tend to use centralised tx submission methods, which creates a chokepoint that will attract censorship)
 - running a plain bitcoind should get you block templates within 90%-99% of optimal, depending on the resources you can allocate to your node
   - (otherwise miners will outsource block template construction to proprietary services in order to be competitive, which again creates a censorship chokepoint)
 - mempool/relay behaviour should be as simple to understand as possible -- for wallet designers, bitcoin users, node operators and miners
   - (otherwise bitcoin will only be usable in theory, not in practice)

I think most of the "incentive compatible" things above aren't really making anything more incentive compatible (miners can just raise their mempool limits to keep the transactions; non-mining nodes aren't receiving any incentives in the first place) but rather making mempool behaviour easier to understand (eg "we removed this tx because there was a race condition when loading the mempool" vs "we removed this tx because it wouldn't be included in any of the next 100 optimal blocks, even if no new txs are submitted").

**Comment by ariard** [comment]
If the state-of-art implementation of BIP331 can be linked here or in the BIP (Like there is https://github.com/glozow/bitcoin/pull/8 and https://github.com/glozow/bitcoin/pull/9) ?

From a reviewer perspective, ideally it’s better to have the proposed specification changes, the proposed Core code changes and the backend code of broadcaster client (e.g Lightning) to grasp a correct mental model of all the interactions.

**Comment by glozow** [comment]
> If the state-of-art implementation of BIP331 can be linked here or in the BIP (Like there is https://github.com/glozow/bitcoin/pull/8 and https://github.com/glozow/bitcoin/pull/9) ?
> From a reviewer perspective, ideally it’s better to have the proposed specification changes, the proposed Core code changes and the backend code of broadcaster client (e.g Lightning) to grasp a correct mental model of all the interactions.

Forgot to respond to this earlier - #27742 is the branch you're looking for :)

**Comment by AndriiHlukhovskyi** [comment]
Super

**Comment by ariard** [comment]
From a reviewing standpoint on #27742, I think it would benefit if the release aim for current package version (BIP331 + nversion=3 policy regime) is explicitly bounded to Lightning time-sensitive confirmation flows, especially in matters of DoS protection:
- broadcast of revoked transaction output spend
- broadcast of a commitment transaction with pending HTLC outputs and its associated second-stage transactions

I think this covers the most sensitive Lightning flows impacted by the pinning attacks mentioned as a design goal by BIP331. However this left of out considerations the following multi-party contract protocols and subsidiary Lightning flows:
- submarine swaps e.g peerswap
- LN dual-funding and splicing (package relay has been discussed as a griefing mitigation on the lightning dev mailing list iirc)
- Taprooted wallet with time-sensitive script path

I think this is opportune to say so on the mailing list and therefore avoid multi-party applications or contracting protocols having to complaint than this deployment of package relay doesn’t fit their use-case (kind of like we have with the mempoolfullrbf debate). Beyond, I would say it’s valuable to announce latest BIP331 (minor) changes, if it does change something for the p2p network, that we would have missed from the Bitcoin Core-side.

**Comment by mehran636311** [comment]
Pellse full item my account open and stat

**Comment by AndriiHlukhovskyi** [comment]
Спасибо огромное очень познавательно я думаю что я тоже что то пропустил
!!! 🔥

чт, 17 авг. 2023 г., 12:58 mehran636311 ***@***.***>:

**Comment by glozow** [comment]
Status update: I've put all the p2p PRs in draft to focus on getting the mempool/validation things done (I was struggling to context-switch and iterate quickly with the 5+ PRs). The v3 stuff also can't make progress until the `AcceptPackage` interface is more stable. After #26711 and more fuzzing, it also might be safe to open submitpackage on mainnet.

I'll continue working on the p2p code and write a lot more tests before opening them again. Please review the mempool stuff if you want to help with the project.

**Comment by tolex469** [comment]
Add new features

**Comment by achow101** [comment]
#26711 was closed. Could we get a longer explanation of why and what the plan is going forward? My understanding is that the plan is to do a limited package relay carveout for just CPFP, then cluster mempool, then the rest of package relay?

**Comment by instagibbs** [comment]
I'll update Ephemeral Anchors PR/BIP once current set of PRs are merged, since I build on https://github.com/bitcoin/bitcoin/pull/28848 for testing

**Comment by glozow** [comment]
There have been discussions on what package validation and RBF rules will look like in a post-cluster mempool world. #26711 either doesn't work with cluster mempool or should not be considered until after cluster mempool. #26711 supports certain types of package submission that we might not be able to continue to support with cluster mempool depending on what approach is taken. There are different ideas on how to consider subsets of packages to accept when it doesn't make sense to take all or none. IIUC the approach taken in #26711 (which is to do subpackage evaluation) no longer has conceptual support which is why I closed it.

Another issue is a circular dependency between cluster mempool and package relay. (1) We want cluster mempool before package RBF since it allows us to assess incentive compatibility much more easily. (2) Cluster mempool requires the removal of CPFP carve out in order to switch from ancestor/descendant-based package limits to cluster limits. But there are people who rely on CPFP carve out for bumping presigned transactions; we need to build a replacement (i.e. package relay + package RBF).

So the new plan of attack is to first create 1-parent-1-child package relay, including package RBF for v3 transactions. This allows the users of CPFP carve out to switch, thus allowing us to remove CPFP carve out and then add cluster mempool. This also allows us to defer the decision-making on what package evaluation and package RBF will look like in a post-cluster mempool world.

**Comment by glozow** [comment]
The OP has been updated to reflect the new plan.

**Comment by glozow** [comment]
Update similar to my irc update: there are a few possible next steps including (1) rebasing #28031 to start adding the orphan resolution module (2) making `TxDownloadManager` internally thread-safe (3) finishing up the one_honest_peer fuzzer.

Nothing is open for review right now as I am a bit busy with other things and still deciding what to do next.

**Comment by jsarenik** [comment]
Let me note here that I noticed the zero (tested with LN Anchor) outputs are not being cleared in a read-only descriptor wallet even after they've been spent by a fee-paying transaction in a package.

UPDATE: no issue because I am unable to reproduce it on current master (6ca6f3b37b99). Seems to have been fixed some time ago.

UPDATE2: See https://github.com/bitcoin/bitcoin/issues/33265 which works on current master (6ca6f3b37b99)

**Comment by instagibbs** [comment]
@jsarenik could you open a new issue?

**Comment by glozow** [comment]
Closing this as completed. We have package validation and relay for the desired use cases; TRUC and cluster mempool fix pinning problems to a great extent.

Here is a list of the most wortwhile ideas to be explored and deployed if deemed necessary (also discussed at [coredev](https://btctranscripts.com/bitcoin-core-dev-tech/2025-10/package-relay)):

(1) Extended topologies for TRUC-like things. For example, chains longer than 2 (the topology restriction should require a branching factor of 1 at each generation) can help with tree-like L2 structures (source: requests from various developers) The child size limit should be distributed amongst the descendants to maintain pinning bounds.

(2) Sender-initiated package relay can be a way to increase orphan resolution throughput and save bandwidth; it should be carefully measured in real-world scenarios. This may not be very impactful given the current transaction volume. See #33500

(3) Arbitrary package acceptance. This problem (namely, more generalized package RBF) is not as tractable as one would hope even with cluster mempool, and perhaps not necessary unless (4) is desired. There is a writeup somewhere and several draft implementations on my repository.

(4) Ancestor package relay (BIP 331) to eliminate txid-based transaction relay. There should be some discussion as to whether this benefit warrants a protocol change. It may also be best to do something closer to "cluster prefix" instead of ancestor set so that fee-bumping nieces/nephews are included. Requires (3) unless we are ok with simply submitting one by one.
